File: string/67.c
class Solution:
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        ret = ''
        flag = 0
        l1 = len(a)
        l2 = len(b)
        index1 = l1-1
        index2 = l2-1
        while index1 >= 0 and index2 >= 0:
            s = int(a[index1]) + int(b[index2]) + flag
            c = 1 if s == 1 or s == 3 else 0
            flag = 1 if s == 2 or s == 3 else 0
            ret = str(c) + ret
            index1 -= 1
            index2 -= 1
            
        
        while index1 >= 0:
            s = int(a[index1]) + flag
            c = 1 if s == 1 else 0
            flag = 1 if s == 2 else 0            
            ret = str(c) + ret
            index1 -= 1
            
        while index2 >= 0:
            s = int(b[index2]) + flag
            c = 1 if s == 1 else 0
            flag = 1 if s == 2 else 0            
            ret = str(c) + ret
            index2 -= 1            
            
        if flag == 1:
            ret = '1' + ret
            
        return ret
            
        

File: string/67.go
func addBinary(a string, b string) string {
    var flag bool
    var ret string
    var i, j int
    for i, j = len(a)-1, len(b)-1; i >= 0 && j >= 0; i, j = i-1, j-1 {
        if flag {
            if a[i] == '1' && b[j] == '1' {
                ret = "1" + ret
            }else if a[i] == '0' && b[j] == '0' {
                ret = "1" + ret
                flag = false
            }else {
                ret = "0" + ret
            }
        }else {
            if a[i] == '1' && b[j] == '1' {
                ret = "0" + ret
                flag = true
            }else if a[i] == '0' && b[j] == '0' {
                ret = "0" + ret
            }else {
                ret = "1" + ret
            }            
        }
    }
    for i >= 0 {
        if flag {
            if a[i] == '1' {
                ret = "0" + ret
            }else {
                ret = "1" + ret
                flag = false
            }
        }else {
            if a[i] == '1' {
                ret = "1" + ret
            }else {
                ret = "0" + ret
            }
        }
        i--
    }
    for j >= 0 {
        if flag {
            if b[j] == '1' {
                ret = "0" + ret
            }else {
                ret = "1" + ret
                flag = false
            }
        }else {
            if b[j] == '1' {
                ret = "1" + ret
            }else {
                ret = "0" + ret
            }
        }
        j--
    }    
    if flag {
        ret = "1" + ret
    }
    return ret
}
